crates/bifrost/src/loglet/loglet_tests.rs

```diff
@@ -393,3 +399,150 @@ pub async fn single_loglet_readstream_test_with_trims(
+        assert!(!tail.is_sealed());
+    }
+
+    // append 5 records. Offsets [1..5]
+    for i in 1..=5 {
+        loglet.append(Bytes::from(format!("record{}", i))).await?;
+    }
+
+    loglet.seal().await?;
+
+    // attempt to append 5 records. Offsets [6..10]. Expected to fail since seal happened on the same client.
+    for i in 6..=10 {
+        let res = loglet.append(Bytes::from(format!("record{}", i))).await;
+        assert_that!(res, err(pat!(AppendError::Sealed)));
+    }
+
+    let tail = loglet.find_tail().await?;
+    // Seal must be applied after commit index 5 since it has been acknowledged (tail is 6 or higher)
+    assert_that!(tail, pat!(TailState::Sealed(gt(LogletOffset::from(5)))));
+
+    Ok(())
+}
+
+/// Validates that appends fail after find_tail() returned Sealed()
+pub async fn loglet_test_append_after_seal_concurrent(
+    loglet: Arc<dyn Loglet>,
+) -> googletest::Result<()> {
+    use futures::TryStreamExt as _;
+
+    setup();
+
+    assert_eq!(None, loglet.get_trim_point().await?);
+    {
+        let tail = loglet.find_tail().await?;
+        assert_eq!(LogletOffset::OLDEST, tail.offset());
+        assert!(!tail.is_sealed());
+    }
+    let warmup_appends = 1000;
+    let concurrent_appenders = 20;
+    // +1 for the main task waiting on all concurrent appenders
+    let append_barrier = Arc::new(Barrier::new(concurrent_appenders + 1));
+
+    let mut appenders: JoinSet<googletest::Result<_>> = JoinSet::new();
+    for appender_id in 0..concurrent_appenders {
+        appenders.spawn({
+            let loglet = loglet.clone();
+            let append_barrier = append_barrier.clone();
+            async move {
+                let mut i = 1;
+                let mut committed = Vec::new();
+                let mut warmup = true;
+                loop {
+                    let res = loglet
+                        .append(Bytes::from(format!("appender-{}-record{}", appender_id, i)))
+                        .await;
+                    i += 1;
+                    if i > warmup_appends && warmup {
+                        println!("appender({}) - warmup complete....", appender_id);
+                        append_barrier.wait().await;
+                        warmup = false;
+                    }
+                    match res {
+                        Ok(offset) => {
+                            committed.push(offset);
+                        }
+                        Err(AppendError::Sealed) => {
+                            break;
+                        }
+                        Err(e) => fail!("unexpected error: {}", e)?,
+                    }
+                    // give a chance to other tasks to work
+                    tokio::task::yield_now().await;
+                }
+                Ok(committed)
+            }
+        });
+    }
+
+    // Wait for some warmup appends
+    println!(
+        "Awaiting all appenders to reach at least {} appends",
+        warmup_appends
+    );
```

Comment:
Are those printlns intentionally left in the code? If yes, then we could maybe replace them with logging statements which would allow to control them via the config level.

Reply:
Yes. Those are left intentionally, very helpful when running the tests with `--no-capture` and hidden otherwise. I'm not sure if tracing/logging with levels add value to be honest but happy to change it if you think otherwise.

Reply:
I guess it is ok since if the test passes then we don't see the log output.